### PR TITLE
Defined `terraform-cloud` group of commands

### PR DESCRIPTION
**Pull request by neomatamune**
Those commands allow the interaction with the API for the web version of terraform

Commands interact with workspaces :
- Create/Read/Delete operations are available for workspaces
- Run of a workspace is available
- Outputs and Last Run status can be queried
- CRUD for workspace variables

Information on setup : https://github.com/Cosmo-Tech/Babylon/pull/32#issuecomment-1285470668
List of tests : https://github.com/Cosmo-Tech/Babylon/pull/32#issuecomment-1285455881

### CODE REVIEW

**Comment by neomatamune** [comment]
# List of commands and associated tests

```bash
# Creation of the test workspace, -s option should set it's id as a selected ID
babylon terraform-cloud workspace create TestWorkspace -s
# Test of creation with an existing workspace name : should display an error
babylon terraform-cloud workspace create TestWorkspace -s
# Test the get : should return the workspace previously created
babylon terraform-cloud workspace get
# Test a non existing id : should display an error (workspace ids on the API are of the form "ws-<hash>" so any other string should work as non existing)
babylon terraform-cloud workspace get -w NonExistingWorkspace
# Test get-all : should return a long list of workspaces
babylon terraform-cloud workspace get-all
# Test of outputs : should return an error for our workspace as no outputs exists for it
babylon terraform-cloud workspace outputs
# Test creation of a run : should return information on the created run
babylon terraform-cloud workspace run
# Test getting the last run : should return information close to the one of the previous commands (but with a different state)
babylon terraform-cloud workspace last-run
# Test getting all variables : should inform that no vars exists
babylon terraform-cloud workspace vars get-all
# Test creating a variable : the var should be created and available
babylon terraform-cloud workspace vars create MyVar MyValue MyDescription
# Test creating a HCL variable : the var should be created and available
babylon terraform-cloud workspace vars create MyHCLVar MyValue MyDescription --hcl
# Test creating a sensitive variable : the var should be created and available
babylon terraform-cloud workspace vars create MySensitiveVar MyValue MyDescription --sensitive
# Test creating a var already defined : should display an error
babylon terraform-cloud workspace vars create MyVar MyValue MyDescription
# Test getting an existing var : should return the var infos
babylon terraform-cloud workspace vars get MyVar
# Test getting an existing sensitive var : should return the var infos, but 'attirbutes.value' should be at None
babylon terraform-cloud workspace vars get MySensitiveVar
# Test getting a non-existing var : should display an error
babylon terraform-cloud workspace vars get MyNonExistingVar
# Test updating a var : should display an update of the var
babylon terraform-cloud workspace vars update MyVar --value MyUpdatedValue --description MyUpdatedDescription
# Test listing existing vars : now should display a list of vars
babylon terraform-cloud workspace vars get-all
# Test deleting a var : should display that the var is being deleted
babylon terraform-cloud workspace vars delete MyVar
# Test deleting a non existing var : should display an error
babylon terraform-cloud workspace vars delete MyNonExistingVar
# Test getting a deleted var : should display an error
babylon terraform-cloud workspace vars get MyVar
# Test deleting a workspace : should ask confirmation and then delete the workspace
babylon terraform-cloud workspace delete
# Test deleting a non existing workspace : should display an error
babylon terraform-cloud workspace vars delete -w NonExistingWorkspace
```

**Comment by neomatamune** [comment]
# Setting up the working dir for Terraform cloud
Update your working dir by using the command `babylon working-dir complete`
You should now have 2 new files : `terraform_cloud.yaml` and `terraform_workspace.yaml`

## `terraform_cloud.yaml`
You need to fill this file with information from the terraform wep application : 
- `url` : the url to the terraform web app you want to use
- `organization` :  the name of the organization you use inside the web app
- `token` : an API access token generated inside the application (more info here : https://app.terraform.io/app/settings/tokens )

## `terraform_workspace.yaml`
This file will contain information about the specific workspace you want to control, the following information can be filled to allow the creation of the workspace :
- `vcs_identifier` : the identifier of the repository used to create the workspace
- `vcs_branch`: the branch of the repository to use for the workspace
- `vcs_auth_token_id`: the id of a oauth2 token generated inside the terraform web app and linked to the github webapp (more information can be found on `https://app.terraform.io/app/[organization]/settings/version-control` )
- `working_directory` : the directory inside the repository that will be used for the workspace

The following information can be added manually or can be filled automatically by adding the `--select` option during the creation of a workspace :
- `workspace_id` : the id of the workspace to use by default in subsequent commands

File: Babylon/groups/terraform_cloud/groups/workspace/groups/vars/list_all_vars.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    workspace_name = ws['data']['attributes']['name']
+    ws_vars = api.vars.list(workspace_name=workspace_name)
+    r = []
+
+    existing_keys = []
+
+    for ws_var in sorted(ws_vars.get('data'), key=lambda _r: _r.get('attributes', {}).get('key', '')):
+        r.append(ws_var)
+        existing_keys.append(ws_var['attributes']['key'])
+
```
Comment (o-reo) [inline Babylon/groups/terraform_cloud/groups/workspace/groups/vars/list_all_vars.py:35]: I would define the lambda function before, I already answered this with an example before.

File: Babylon/groups/terraform_cloud/groups/workspace/groups/vars/list_all_vars.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    if lookup_var_sets:
+        var_sets = api.var_sets.list_for_workspace(workspace_id)
+
+        for _var_set in var_sets['data']:
+            for vs_var in api.var_sets.list_vars_in_varset(_var_set['id'])['data']:
+                _k = vs_var['attributes']['key']
+                if _k not in existing_keys:
+                    r.append(vs_var)
```
Comment (o-reo) [inline Babylon/groups/terraform_cloud/groups/workspace/groups/vars/list_all_vars.py:43]: you can try a list comprehension here as well, but I am not sure of its readability, you tell me
```python
return [
   vs_var["attributes"]["key"]
   for vs_var in api.var_sets.list_vars_in_varset(_var_set['id'])['data'])
   for _var_set in var_sets["data"]
   if _vs_var["attributes"]["key"] not in existing_keys
]
```

File: Babylon/utils/working_dir.py [hunk 3]
```diff
@@ -121,6 +122,17 @@ def get_yaml_key(self, yaml_path: str, yaml_key: str) -> Optional[object]:
         except:
```
Comment (o-reo) [inline Babylon/utils/working_dir.py:122]: Bare `except` would not pass with flake8

**Review by o-reo — COMMENTED** [review]
I made some suggestions, I think using open outside of a with context can break
I still have to test the commands though

File: Babylon/groups/azure/__init__.py [hunk 4]
```diff
@@ -1,11 +1,13 @@
...
+
 from azure.core.exceptions import ClientAuthenticationError
-from click import group, pass_context
+from azure.identity import DefaultAzureCredential
+from click import group
+from click import pass_context
 from click.core import Context
-from Babylon.utils.decorators import require_platform_key
 
+from Babylon.utils.decorators import require_platform_key
```
Comment (o-reo) [inline Babylon/groups/azure/__init__.py:10]:
```suggestion
from .....utils.decorators import require_platform_key
```

File: tests/resources/workingdir/terraform_workspace.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# workspace_id: id of the workspace to use by default
+workspace_id: ""
+# working_directory: working directory of the workspace
+working_directory: ""
+# vcs_identifier: identifier of the vcs associated to the workspace
+vcs_identifier: ""
+# vcs_branch: branch in the vcs used by the workspace
+vcs_branch: ""
+# vcs_oauth_token_id: id to the oauth token used to connect to the vcs
+vcs_oauth_token_id: ""
```
Comment (o-reo) [inline tests/resources/workingdir/terraform_workspace.yaml:10]:
```suggestion
vcs_oauth_token_id: ""

```
Missing empty line

**Review by o-reo — COMMENTED** [review]
LGTM, you can add suggestion to batch commit and commit all at once

**Pull request merged**